extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -1252,17 +1250,22 @@ export class AmpA4A extends AMP.BaseElement {
             'height': this.element.getAttribute('height'),
             'width': this.element.getAttribute('width'),
             'src': srcPath,
-            'name': nameData,
           }, SHARED_IFRAME_PROPERTIES));
-      if (method == XORIGIN_MODE.NAMEFRAME) {
         // TODO(bradfrizzell): change name of function and var
-        const attributes = getContextMetadata(
-            this.win, this.element, this.sentinel);
-        attributes['creative'] = creative;
-        const name = JSON.stringify(attributes);
-        // Need to reassign the name once we've generated the context
-        // attributes off of the iframe. Need the iframe to generate.
+      let contextMetadata = getContextMetadata(
+          this.win, this.element, this.sentinel);
+      if (method == XORIGIN_MODE.NAMEFRAME) {
+        contextMetadata['creative'] = creative;
+        name = JSON.stringify(contextMetadata);
+      } else if (method == XORIGIN_MODE.SAFEFRAME) {
+        contextMetadata = JSON.stringify(contextMetadata);
+        name = `${SAFEFRAME_VERSION};${creative.length};${creative}` +
+            `${contextMetadata}`;
+      }
+      if (name) {
         iframe.setAttribute('name', name);
+      }
+      if (this.sentinel) {
         iframe.setAttribute('data-amp-3p-sentinel', this.sentinel);
       }
       return this.iframeRenderHelper_(iframe);
```

Comment:
You could move setting of name and sentinel into iframeRenderHelper_ which will ensure its always set when using xdomain method.